src/resolution/composite-arns-resolver.ts

```diff
@@ -57,21 +70,189 @@ export class CompositeArNSResolver implements NameResolver {
+      const span = tracer.startSpan(
+        'CompositeArNSResolver.resolveWithResolver.resolutionPromise',
+        {
+          attributes: {
+            arnsName: name,
+            resolver: resolver.constructor.name,
+          },
+        },
+        trace.setSpan(context.active(), parentSpan),
+      );
+      try {
+        const resolution = await resolver.resolve({
+          name,
+          baseArNSRecordFn,
+          signal,
+        });
+
+        // Only cache and return resolutions where resolvedAt is set
+        if (resolution.resolvedAt !== undefined) {
+          this.resolutionCache.set(
+            name,
+            Buffer.from(JSON.stringify(resolution)),
+          );
+
+          metrics.arnsResolutionResolverCount.inc({
+            resolver: resolver.constructor.name,
+          });
+
+          this.log.verbose('Resolved name', {
+            name,
+            resolution,
+            resolvedBy: resolver.constructor.name,
+          });
+
+          return resolution;
+        }
+
+        return undefined;
+      } catch (e: any) {
+        const error = e as Error;
+        this.log.error('Error resolving name with resolver', {
+          name,
+          resolver: resolver.constructor.name,
+          message: error.message,
+          stack: error.stack,
+        });
+        span.recordException(error);
+        return undefined;
+      } finally {
+        span.end();
+      }
+    })();
+
+    if (isLastResolver) {
+      return resolutionPromise;
+    }
+
+    return pTimeout(resolutionPromise, {
+      milliseconds: this.resolverTimeoutMs,
+      signal,
     });
   }
 
-  async resolve({ name }: { name: string }): Promise<NameResolution> {
+  private async resolveParallel({
+    name,
+    baseArNSRecordFn,
+    signal,
+  }: {
+    name: string;
+    baseArNSRecordFn: () => Promise<any>;
+    signal: AbortSignal;
+  }): Promise<NameResolution | undefined> {
+    const span = tracer.startSpan('CompositeArNSResolver.resolverParallel', {
+      attributes: {
+        arnsName: name,
+      },
+    });
+
+    // Return the pending resolution for the name if it exists
+    if (this.pendingResolutions[name]) return this.pendingResolutions[name];
+
+    const limit = pLimit(this.maxConcurrentResolutions);
+    const alreadyResolvedAbort = new AbortController();
+    const resolutionPromises = this.resolvers.map((resolver, index) => {
+      const isLastResolver = index === this.resolvers.length - 1;
+      return limit(async () => {
+        // Check if already aborted
+        if (alreadyResolvedAbort.signal.aborted) return undefined;
+
+        const resolution = await this.resolveWithResolver({
+          resolver,
+          name,
+          baseArNSRecordFn,
+          isLastResolver,
+          signal: anySignal([signal, alreadyResolvedAbort.signal]),
+          parentSpan: span,
+        });
+
+        if (resolution) {
+          // Signal other resolvers to abort when we find a valid resolution
```

Comment:
ah! disregard previous comment - this is what was needed to get accurate metrics around resolver counts